[Step3] 로또(자동) PR 요청 드립니다.

90mansik opened the pull request:
민제님 안녕하세요!
최대한 규칙에 맞게 해보려고 노력했습니다.

통계, 로또 검증하는 부분이 미흡한 것 같습니다.
피드백 주시면 열심히 고쳐 보겠습니다!!

감사합니다.

Commit:
feat: Moeny 도메인 생성

- 입력받은 숫자를 횟수로 반환 기능 구현
- 입력받은 금액이 0원 이하일 경우 Exception 발생 기능 구현
- 입력받은 금액 천원 단위가 아닐 경우 Exception 발생 기능 구현

Commit:
docs: README.md 로또 요구사항 정리

Commit:
test: Lotto 도메인 테스트 구현

- 로또번호 6개 저장 테스트
- 로또번호 중복 입력시 Exception 발생 검증
- 로또번호 6자리와 다를 경우 Exception 발생 검증

Commit:
feat: Lotto번호를 저장하는 기능 추가

- 로또번호 6개 저장 기능 구현
- 로또번호 유효성 검증 기능 구현(중복,자리수 검증)

Commit:
test: README.md 완료항목 체크

Commit:
feat: 로또번호 유효성 체크를 위한 도메인 구현

Commit:
test: 로또번호 생성 테스트 구현

Commit:
feat: 자동 로또번호 생성 기능 구현

Commit:
feat: Lotto클래스 lottoNumbers 제네릭스 값 수정

- AS-IS : List<Integer>
- TO-BE : List<LottoNumber>

Commit:
style: Lotto, LottoTest 변수명 수정

Commit:
feat: LottoNumber equlas,hasCode,toString 메서드 추가

Commit:
feat:  로또 구입 기능 추가

- 코드 라인 정리
- 로또 구입 기능 추

Commit:
test: 로또 등수 검증 테스트 구현

Commit:
feat: 로또 등수 체그 기능 구현

Commit:
feat: 로또 일치값을 저장하는 WinReport 클래스 구현

Commit:
feat: input,output view 구현

Commit:
feat: LottoMain 기능 구현

Commit:
test: README.md 완료항목 체크

minzzang commented on README.md:
```diff
@@ -22,5 +22,31 @@
+> - [x] 5.통계
+>   -  [x] 6.1 당첨번호와 로또번호 비교 카운트 반환
+>   -  [x] 6.2 당첨 통계를 이용해서 수익률 계산
+>   
```
상세 기능 정리 꼼꼼하게 잘 하셨네요 👍 👍

minzzang commented on src/main/java/domain/Money.java:
```diff
@@ -0,0 +1,33 @@
+    private static final int DEFAULT_LOTTO_PRICE = 1000;
+    private static final int MIN_MONEY = 0;
+
+    private final int money;
```
Money클래스를 불변으로 잘 만드셨네요~ 좋습니다 👍

minzzang commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,45 @@
+public class LottoNumber {
+
+    private static final int LOTTO_MIN_NUMBER = 1;
+    private static final int LOTTO_MAX_NUMBER = 45;
```
전체적으로 컨벤션이 일치하지 않는 느낌이네요! 클래스 선언 후 한줄 띄고 변수를 선언한 부분도 있고 띄지 않고 선언한 부분이 있는데 통일이 되면 더 좋을 것 같아요~ 그리고 스태틱 변수와 인스턴스 변수는 한 줄 띄면 보기 더 좋을 것 같습니다!!

minzzang commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,57 @@
+
+public class Lotto {
+    private static final int LOTTO_SIZE = 6;
+    private final List<LottoNumber> lottoNumbers;
```
일급 컬렉션 활용 좋습니다 ~ 👍

minzzang commented on src/main/java/domain/AutoLottoMachine.java:
```diff
@@ -0,0 +1,38 @@
+        setupLotto();
+    }
+
+    public static void setupLotto(){
```
이 부분을 public static으로 한 이유가 있을까요? 외부에서 호출하면 중복으로 로또 넘버가 들어갈 수 있을 것 같네요!

minzzang commented on src/main/java/domain/AutoLottoMachine.java:
```diff
@@ -0,0 +1,38 @@
+
+
+    @Override
+    public List<Lotto> purchaseLotte(int lottoTicketCount) {
```
lottoList를 일급 컬렉션으로 만들어 보는건 어떨까요~?
그리고 변수명에 자료구조를 포함하는 것보다 lottoList -> lottos로 하는건 어떤가요~?
List에서 다른 컬렉션으로 변경했을 경우 어색해질 수도 있을 것 같아요 :)

minzzang commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,57 @@
+        }
+    }
+
+    public List<LottoNumber> getLottoNumbers(){
```
일급 컬렉션을 그대로 반환하면 어떤 문제가 생길 수 있을까요?
https://tecoble.techcourse.co.kr/post/2021-04-26-defensive-copy-vs-unmodifiable/

minzzang commented on src/main/java/domain/WinLotto.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    private int containNumbers(LottoNumber lottoNumber) {
+        if(winLottoNumbers.getLottoNumbers().contains(lottoNumber)){
```
lottoNumbers를 가져와서 사용하는 것보다 lotto객체에 메시지를 보내는 건 어떨까요~? 
https://tecoble.techcourse.co.kr/post/2020-06-02-law-of-demeter/

minzzang commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,56 @@
+
+public class OutputView {
+
+    private static final int[] PRIZE_MONEY = {0, 0, 0, 5000, 50000, 1500000, 2000000000};
```
enum을 활용해봐도 좋을 것 같아요! 😄

minzzang commented on src/test/java/study/step3/LottoTest.java:
```diff
@@ -0,0 +1,69 @@
+    @Test
+    @DisplayName("로또번호 6개를 저장한다.")
+    public void 로또번호_저장(){
+        lottoNumbers = new ArrayList<>();
```
테스트 픽스처를 활용하면 중복 코드를 줄일 수 있을 것 같아요~
https://thalals.tistory.com/375

minzzang commented on src/test/java/study/step3/WinLottoTest.java:
```diff
@@ -0,0 +1,84 @@
+        lotto = new Lotto(lottoNumbers);
+    }
+    @Test
+    @DisplayName("로또번호 3개 맞춘 카운트 검증")
```
로또번호 n개 맞춘 검증을 따로 만들기 보다 ParameterizedTest를 활용해 하나로 합치는건 어떨까요~?

minzzang commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,56 @@
+    public static void outputProfit(WinLotto winLotto, int lottoTicketCount){
+        WinReport winReport = winLotto.getWinLottoReport();
+        double profit = 0;
+        for(int i= WIN_MIN_NUMBER; i<PRIZE_MONEY.length; i++){
```
winReport에 우승 로또 정보가 있으니 수익률 계산에 대한 책임을 위임할 수 있을 거 같은데 어떻게 생각하시나요~?

minzzang reviewed (changes requested):
안녕하세요 지용님~!
3단계 미션 아주 잘 수행해주셨습니다~

규칙에 맞게 작성하려는게 코드에 보였고 전체적으로
코드가 잘 읽히네요!! :+1:

다만 수정하면 더 좋아질 부분이 있어서 코멘트 남겼는데
확인 부탁드립니다~ :pray: 
애매하거나 이해하기 어려운 부분은 편하게 DM보내주세요~

그럼 오늘 하루도 화이팅입니다. :smile:

Commit:
fix: 불필요한 public static 사용 제거

AS IS:
 public static void setupLotto()
 : 외부에서 호출하면 중복으로 로또 넘버가 추가될 수 있음.
TO BE:
 private void setUpLotto()

Commit:
style: 전체 코드 작성 규칙 통일화

Commit:
refactor: 테스트 중복 코드 제거

- ParameterizedTest를 통한 중복 코드 제거

Commit:
refactor: 불변객체를 만들기 위한 get 메서드 수정

Commit:
feat: Lotts 일급컬렉션 생성

Commit:
refactor: Lotts 일급컬렉션 사용하도록 변경

Commit:
feat: 상금의 맞힌 갯수, 상금을 저장 하기 위한 enum 생성

Commit:
refactor: 수익율계산 객체 책임 분리, enum 활용하도록 수정

- 수익률의 계산 책임 분리
  AS IS : OutputView에서 계산
  To BE : WinReport에서 계산
- PrizeMoney enum을 사용하도록 리팩터링

Commit:
refactor: 테스트 픽스처를 통한 중복 소스 제거

minzzang commented on src/main/java/domain/AutoLottoMachine.java:
```diff
@@ -0,0 +1,38 @@
+
+    private void setupLotto() {
+        for (int i = LOTTE_MIN_NUMBER; i <= LOTTE_MAX_NUMBER; i++) {
+            LOTTO_NUMBERS.add(new LottoNumber(i));
```
다음 미션 진행하면서 로또 번호를 계속 생성하는 것이 아닌 캐싱해서 재 사용해보는 것도 고민해보면 좋을 것 같아요~

90mansik replied:
넵 감사합니다! step4 진행하면서 적용 하겠습니다!.
꼼꼼한 피드백 감사드립니다!

minzzang reviewed (approved):
안녕하세요 지용님~!
피드백 잘 반영해주셨습니다!! :100:

테스트 코드도 깔끔해지고 객체의 역할 분리도 더 잘된 것 같네요.
코멘트 하나 남겼는데 확인 부탁드립니다~ :smile:

그럼 다음 미션도 화이팅입니다!  :muscle:
